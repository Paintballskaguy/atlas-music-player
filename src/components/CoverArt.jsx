import React from "react";
import coverArtImage from '../assets/placeholder.svg';

const CoverArt = () => {
  return (
    <div className="w-48 h-48 aspect-square overflow-hidden rounded-lg">
      <img src={coverArtImage} alt="Cover Art" className="w-full h-full object-cover" />
    </div>
  );
};

export default CoverArt;