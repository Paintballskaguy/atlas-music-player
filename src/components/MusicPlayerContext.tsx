import React, { createContext, useContext, useReducer, useEffect, ReactNode } from 'react';
import ApiService from '../services/api';
import { PlaylistSong, Song, MusicPlayerState, MusicPlayerContextType, RepeatMode } from '../types';

const MusicPlayerContext = createContext<MusicPlayerContextType | undefined>(undefined);

// Action types
const ACTIONS = {
  SET_LOADING: 'SET_LOADING',
  SET_PLAYLIST: 'SET_PLAYLIST',
  SET_CURRENT_SONG: 'SET_CURRENT_SONG',
  SET_CURRENT_SONG_DETAILS: 'SET_CURRENT_SONG_DETAILS',
  SET_LYRICS: 'SET_LYRICS',
  SET_PLAYING: 'SET_PLAYING',
  SET_CURRENT_TIME: 'SET_CURRENT_TIME',
  SET_DURATION: 'SET_DURATION',
  SET_VOLUME: 'SET_VOLUME',
  SET_PLAYBACK_SPEED: 'SET_PLAYBACK_SPEED',
  SET_SHUFFLE: 'SET_SHUFFLE',
  SET_REPEAT: 'SET_REPEAT',
  SET_ERROR: 'SET_ERROR',
  CLEAR_ERROR: 'CLEAR_ERROR'
} as const;

type ActionType = typeof ACTIONS[keyof typeof ACTIONS];

type Action = 
  | { type: 'SET_LOADING'; payload: boolean }
  | { type: 'SET_PLAYLIST'; payload: PlaylistSong[] }
  | { type: 'SET_CURRENT_SONG'; payload: PlaylistSong | null }
  | { type: 'SET_CURRENT_SONG_DETAILS'; payload: Song | null }
  | { type: 'SET_LYRICS'; payload: string | null }
  | { type: 'SET_PLAYING'; payload: boolean }
  | { type: 'SET_CURRENT_TIME'; payload: number }
  | { type: 'SET_DURATION'; payload: number }
  | { type: 'SET_VOLUME'; payload: number }
  | { type: 'SET_PLAYBACK_SPEED'; payload: number }
  | { type: 'SET_SHUFFLE'; payload: boolean }
  | { type: 'SET_REPEAT'; payload: RepeatMode }
  | { type: 'SET_ERROR'; payload: string | null }
  | { type: 'CLEAR_ERROR' };

// Initial state
const initialState: MusicPlayerState = {
  playlist: [],
  currentSong: null,
  currentSongDetails: null,
  lyrics: null,
  isPlaying: false,
  currentTime: 0,
  duration: 0,
  volume: 50,
  playbackSpeed: 1,
  shuffle: false,
  repeat: false,
  isLoading: true,
  error: null
};

// Reducer
function musicPlayerReducer(state: MusicPlayerState, action: Action): MusicPlayerState {
  switch (action.type) {
    case ACTIONS.SET_LOADING:
      return { ...state, isLoading: action.payload };
    case ACTIONS.SET_PLAYLIST:
      return { ...state, playlist: action.payload };
    case ACTIONS.SET_CURRENT_SONG:
      return { ...state, currentSong: action.payload };
    case ACTIONS.SET_CURRENT_SONG_DETAILS:
      return { ...state, currentSongDetails: action.payload };
    case ACTIONS.SET_LYRICS:
      return { ...state, lyrics: action.payload };
    case ACTIONS.SET_PLAYING:
      return { ...state, isPlaying: action.payload };
    case ACTIONS.SET_CURRENT_TIME:
      return { ...state, currentTime: action.payload };
    case ACTIONS.SET_DURATION:
      return { ...state, duration: action.payload };
    case ACTIONS.SET_VOLUME:
      return { ...state, volume: action.payload };
    case ACTIONS.SET_PLAYBACK_SPEED:
      return { ...state, playbackSpeed: action.payload };
    case ACTIONS.SET_SHUFFLE:
      return { ...state, shuffle: action.payload };
    case ACTIONS.SET_REPEAT:
      return { ...state, repeat: action.payload };
    case ACTIONS.SET_ERROR:
      return { ...state, error: action.payload, isLoading: false };
    case ACTIONS.CLEAR_ERROR:
      return { ...state, error: null };
    default:
      return state;
  }
}

interface MusicPlayerProviderProps {
  children: ReactNode;
}

// Provider component
export const MusicPlayerProvider: React.FC<MusicPlayerProviderProps> = ({ children }) => {
  const [state, dispatch] = useReducer(musicPlayerReducer, initialState);

  // Load initial playlist
  useEffect(() => {
    const loadPlaylist = async () => {
      try {
        dispatch({ type: ACTIONS.SET_LOADING, payload: true });
        const playlist = await ApiService.fetchPlaylist();
        dispatch({ type: ACTIONS.SET_PLAYLIST, payload: playlist });
        
        // Set first song as current if playlist exists
        if (playlist && playlist.length > 0) {
          dispatch({ type: ACTIONS.SET_CURRENT_SONG, payload: playlist[0] });
          await loadSongDetails(playlist[0].id);
        }
        
        dispatch({ type: ACTIONS.SET_LOADING, payload: false });
      } catch (error) {
        dispatch({ 
          type: ACTIONS.SET_ERROR, 
          payload: error instanceof Error ? error.message : 'Unknown error occurred'
        });
      }
    };

    loadPlaylist();
  }, []);

  // Actions
  const loadSongDetails = async (songId: string): Promise<void> => {
    try {
      const songDetails = await ApiService.fetchSong(songId);
      dispatch({ type: ACTIONS.SET_CURRENT_SONG_DETAILS, payload: songDetails });
    } catch (error) {
      console.error('Failed to load song details:', error);
    }
  };

  const loadLyrics = async (songId: string): Promise<void> => {
    try {
      const lyricsData = await ApiService.fetchLyrics(songId);
      dispatch({ type: ACTIONS.SET_LYRICS, payload: lyricsData.lyrics });
    } catch (error) {
      console.error('Failed to load lyrics:', error);
    }
  };

  const selectSong = async (song: PlaylistSong): Promise<void> => {
    dispatch({ type: ACTIONS.SET_CURRENT_SONG, payload: song });
    dispatch({ type: ACTIONS.SET_PLAYING, payload: false });
    dispatch({ type: ACTIONS.SET_CURRENT_TIME, payload: 0 });
    await loadSongDetails(song.id);
    await loadLyrics(song.id);
  };

  const togglePlayPause = (): void => {
    dispatch({ type: ACTIONS.SET_PLAYING, payload: !state.isPlaying });
  };

  const getNextSong = (): PlaylistSong | null => {
    if (!state.currentSong || !state.playlist.length) return null;

    if (state.shuffle) {
      const availableSongs = state.playlist.filter(song => song.id !== state.currentSong!.id);
      if (availableSongs.length === 0) return null;
      return availableSongs[Math.floor(Math.random() * availableSongs.length)];
    }

    const currentIndex = state.playlist.findIndex(song => song.id === state.currentSong!.id);
    if (currentIndex < state.playlist.length - 1) {
      return state.playlist[currentIndex + 1];
    }
    return null;
  };

  const nextSong = (): void => {
    const nextSong = getNextSong();
    if (nextSong) {
      selectSong(nextSong);
    }
  };

  const previousSong = (): void => {
    if (!state.currentSong) return;
    
    const currentIndex = state.playlist.findIndex(song => song.id === state.currentSong!.id);
    if (currentIndex > 0) {
      const prevSong = state.playlist[currentIndex - 1];
      selectSong(prevSong);
    }
  };

  const setVolume = (volume: number): void => {
    dispatch({ type: ACTIONS.SET_VOLUME, payload: Math.max(0, Math.min(100, volume)) });
  };

  const setPlaybackSpeed = (speed: number): void => {
    dispatch({ type: ACTIONS.SET_PLAYBACK_SPEED, payload: speed });
  };

  const toggleShuffle = (): void => {
    dispatch({ type: ACTIONS.SET_SHUFFLE, payload: !state.shuffle });
  };

  const toggleRepeat = (): void => {
    const repeatModes: RepeatMode[] = [false, 'one', 'all'];
    const currentIndex = repeatModes.indexOf(state.repeat);
    const nextRepeat = repeatModes[(currentIndex + 1) % repeatModes.length];
    dispatch({ type: ACTIONS.SET_REPEAT, payload: nextRepeat });
  };

  const clearError = (): void => {
    dispatch({ type: ACTIONS.CLEAR_ERROR });
  };

  const contextValue: MusicPlayerContextType = {
    ...state,
    selectSong,
    togglePlayPause,
    nextSong,
    previousSong,
    setVolume,
    setPlaybackSpeed,
    toggleShuffle,
    toggleRepeat,
    loadLyrics,
    clearError
  };

  return (
    <MusicPlayerContext.Provider value={contextValue}>
      {children}
    </MusicPlayerContext.Provider>
  );
};

// Custom hook
export const useMusicPlayer = (): MusicPlayerContextType => {
  const context = useContext(MusicPlayerContext);
  if (!context) {
    throw new Error('useMusicPlayer must be used within a MusicPlayerProvider');
  }
  return context;
};