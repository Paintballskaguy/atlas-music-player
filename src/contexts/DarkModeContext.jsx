import React, { createContext, useContext, useEffect, useState } from 'react';

const DarkModeContext = createContext();

export const useDarkMode = () => {
  const context = useContext(DarkModeContext);
  if (!context) {
    throw new Error('useDarkMode must be used within a DarkModeProvider');
  }
  return context;
};

// Export the provider as a named export
export const DarkModeProvider = ({ children }) => {
  const [isDark, setIsDark] = useState(() => {
    // Initialize from localStorage or system preference
    const saved = localStorage.getItem('darkMode');
    if (saved !== null) return JSON.parse(saved);
    return window.matchMedia('(prefers-color-scheme: dark)').matches;
  });

  useEffect(() => {
    // Save to localStorage whenever it changes
    localStorage.setItem('darkMode', JSON.stringify(isDark));
    
    // Apply dark mode class to html element
    if (isDark) {
      document.documentElement.classList.add('dark');
    } else {
      document.documentElement.classList.remove('dark');
    }
  }, [isDark]);

  useEffect(() => {
    // Listen for system preference changes (optional)
    const mediaQuery = window.matchMedia('(prefers-color-scheme: dark)');
    const handleChange = (e) => {
      // Only update if user hasn't set a preference manually
      const saved = localStorage.getItem('darkMode');
      if (saved === null) {
        setIsDark(e.matches);
      }
    };
    
    mediaQuery.addEventListener('change', handleChange);
    return () => mediaQuery.removeEventListener('change', handleChange);
  }, []);

  const toggleDarkMode = () => {
    setIsDark(prev => !prev);
  };

  const value = {
    isDark,
    toggleDarkMode,
    setIsDark // Optional: provide direct setter
  };

  return (
    <DarkModeContext.Provider value={value}>
      {children}
    </DarkModeContext.Provider>
  );
};

export { DarkModeContext };