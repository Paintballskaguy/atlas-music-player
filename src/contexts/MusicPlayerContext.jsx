import React, { createContext, useContext, useReducer, useEffect } from 'react';
import ApiService from '../services/api';

const MusicPlayerContext = createContext();

// Action types
const ACTIONS = {
  SET_LOADING: 'SET_LOADING',
  SET_PLAYLIST: 'SET_PLAYLIST',
  SET_CURRENT_SONG: 'SET_CURRENT_SONG',
  SET_CURRENT_SONG_DETAILS: 'SET_CURRENT_SONG_DETAILS',
  SET_LYRICS: 'SET_LYRICS',
  SET_PLAYING: 'SET_PLAYING',
  SET_CURRENT_TIME: 'SET_CURRENT_TIME',
  SET_DURATION: 'SET_DURATION',
  SET_VOLUME: 'SET_VOLUME',
  SET_PLAYBACK_SPEED: 'SET_PLAYBACK_SPEED',
  SET_SHUFFLE: 'SET_SHUFFLE',
  SET_REPEAT: 'SET_REPEAT',
  SET_ERROR: 'SET_ERROR',
  CLEAR_ERROR: 'CLEAR_ERROR'
};

const initialState = {
  playlist: [],
  currentSong: null,
  currentSongDetails: null,
  lyrics: null,
  
  // Playback state
  isPlaying: false,
  currentTime: 0,
  duration: 0,
  volume: 50,
  playbackSpeed: 1,
  shuffle: false,
  repeat: false,
  
  // UI state
  isLoading: true,
  error: null
};

// Reducer
function musicPlayerReducer(state, action) {
  switch (action.type) {
    case ACTIONS.SET_LOADING:
      return { ...state, isLoading: action.payload };
    
    case ACTIONS.SET_PLAYLIST:
      return { ...state, playlist: action.payload };
    
    case ACTIONS.SET_CURRENT_SONG:
      return { ...state, currentSong: action.payload };
    
    case ACTIONS.SET_CURRENT_SONG_DETAILS:
      return { ...state, currentSongDetails: action.payload };
    
    case ACTIONS.SET_LYRICS:
      return { ...state, lyrics: action.payload };
    
    case ACTIONS.SET_PLAYING:
      return { ...state, isPlaying: action.payload };
    
    case ACTIONS.SET_CURRENT_TIME:
      return { ...state, currentTime: action.payload };
    
    case ACTIONS.SET_DURATION:
      return { ...state, duration: action.payload };
    
    case ACTIONS.SET_VOLUME:
      return { ...state, volume: action.payload };
    
    case ACTIONS.SET_PLAYBACK_SPEED:
      return { ...state, playbackSpeed: action.payload };
    
    case ACTIONS.SET_SHUFFLE:
      return { ...state, shuffle: action.payload };
    
    case ACTIONS.SET_REPEAT:
      return { ...state, repeat: action.payload };
    
    case ACTIONS.SET_ERROR:
      return { ...state, error: action.payload, isLoading: false };
    
    case ACTIONS.CLEAR_ERROR:
      return { ...state, error: null };
    
    default:
      return state;
  }
}

// Provider component
export const MusicPlayerProvider = ({ children }) => {
  const [state, dispatch] = useReducer(musicPlayerReducer, initialState);

  // Load initial playlist
  useEffect(() => {
    const loadPlaylist = async () => {
      try {
        dispatch({ type: ACTIONS.SET_LOADING, payload: true });
        const playlist = await ApiService.fetchPlaylist();
        dispatch({ type: ACTIONS.SET_PLAYLIST, payload: playlist });
        
        // Set first song as current if playlist exists
        if (playlist && playlist.length > 0) {
          dispatch({ type: ACTIONS.SET_CURRENT_SONG, payload: playlist[0] });
          await loadSongDetails(playlist[0].id);
        }
        
        dispatch({ type: ACTIONS.SET_LOADING, payload: false });
      } catch (error) {
        dispatch({ type: ACTIONS.SET_ERROR, payload: error.message });
      }
    };

    loadPlaylist();
  }, []);

  // Actions
  const loadSongDetails = async (songId) => {
    try {
      const songDetails = await ApiService.fetchSong(songId);
      dispatch({ type: ACTIONS.SET_CURRENT_SONG_DETAILS, payload: songDetails });
    } catch (error) {
      console.error('Failed to load song details:', error);
    }
  };

  const loadLyrics = async (songId) => {
    try {
      const lyricsData = await ApiService.fetchLyrics(songId);
      dispatch({ type: ACTIONS.SET_LYRICS, payload: lyricsData.lyrics });
    } catch (error) {
      console.error('Failed to load lyrics:', error);
    }
  };

  const selectSong = async (song) => {
    dispatch({ type: ACTIONS.SET_CURRENT_SONG, payload: song });
    dispatch({ type: ACTIONS.SET_PLAYING, payload: false });
    dispatch({ type: ACTIONS.SET_CURRENT_TIME, payload: 0 });
    await loadSongDetails(song.id);
    await loadLyrics(song.id);
  };

  const togglePlayPause = () => {
    dispatch({ type: ACTIONS.SET_PLAYING, payload: !state.isPlaying });
  };

  const nextSong = () => {
    const currentIndex = state.playlist.findIndex(song => song.id === state.currentSong?.id);
    if (currentIndex < state.playlist.length - 1) {
      const nextSong = state.playlist[currentIndex + 1];
      selectSong(nextSong);
    }
  };

  const previousSong = () => {
    const currentIndex = state.playlist.findIndex(song => song.id === state.currentSong?.id);
    if (currentIndex > 0) {
      const prevSong = state.playlist[currentIndex - 1];
      selectSong(prevSong);
    }
  };

  const setVolume = (volume) => {
    dispatch({ type: ACTIONS.SET_VOLUME, payload: volume });
  };

  const setPlaybackSpeed = (speed) => {
    dispatch({ type: ACTIONS.SET_PLAYBACK_SPEED, payload: speed });
  };

  const toggleShuffle = () => {
    dispatch({ type: ACTIONS.SET_SHUFFLE, payload: !state.shuffle });
  };

  const toggleRepeat = () => {
    const repeatModes = [false, 'one', 'all'];
    const currentIndex = repeatModes.indexOf(state.repeat);
    const nextRepeat = repeatModes[(currentIndex + 1) % repeatModes.length];
    dispatch({ type: ACTIONS.SET_REPEAT, payload: nextRepeat });
  };

  const contextValue = {
    ...state,
    // Actions
    selectSong,
    togglePlayPause,
    nextSong,
    previousSong,
    setVolume,
    setPlaybackSpeed,
    toggleShuffle,
    toggleRepeat,
    loadLyrics,
    clearError: () => dispatch({ type: ACTIONS.CLEAR_ERROR })
  };

  return (
    <MusicPlayerContext.Provider value={contextValue}>
      {children}
    </MusicPlayerContext.Provider>
  );
};

// Custom hook
export const useMusicPlayer = () => {
  const context = useContext(MusicPlayerContext);
  if (!context) {
    throw new Error('useMusicPlayer must be used within a MusicPlayerProvider');
  }
  return context;
};