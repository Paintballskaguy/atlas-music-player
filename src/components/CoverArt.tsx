import React, { useState } from "react";
import { useMusicPlayer } from '../contexts/MusicPlayerContext';
import coverArtImage from '../assets/placeholder.svg';

interface CoverArtProps {
  src?: string | null;
  alt?: string;
}

const CoverArt: React.FC<CoverArtProps> = ({ alt = "Cover Art" }) => {
  const { currentSongDetails, lyrics, loadLyrics, currentSong } = useMusicPlayer();
  const [isHovered, setIsHovered] = useState<boolean>(false);

  const handleMouseEnter = async (): Promise<void> => {
    setIsHovered(true);
    if (currentSong && !lyrics) {
      await loadLyrics(currentSong.id);
    }
  };

  const handleMouseLeave = (): void => {
    setIsHovered(false);
  };

  const coverSrc = currentSongDetails?.cover;

  return (
    <div 
      className="w-full aspect-square bg-gradient-to-br from-custom-teal-100 to-custom-blue-200 dark:from-custom-teal-800 dark:to-custom-blue-900 rounded-3xl overflow-hidden flex items-center justify-center shadow-2xl border-4 border-custom-yellow-200 dark:border-custom-yellow-600 transition-all duration-300 relative cursor-pointer"
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
    >
      {isHovered && lyrics ? (
        <div className="absolute inset-0 bg-black bg-opacity-80 flex items-center justify-center p-4 z-10">
          <div className="text-white text-center max-h-full overflow-y-auto">
            <h3 className="font-bold text-lg mb-2">Lyrics</h3>
            <div className="text-sm whitespace-pre-line">
              {lyrics}
            </div>
          </div>
        </div>
      ) : null}
      
      <div className="w-full h-full flex items-center justify-center">
        {coverSrc ? (
          <img 
            src={coverSrc} 
            alt={alt} 
            className="w-full h-full object-cover" 
          />
        ) : (
          <img 
            src={coverArtImage} 
            alt="Placeholder cover art" 
            className="w-20 h-20 opacity-30 dark:opacity-50 filter drop-shadow-lg"
          />
        )}
      </div>
    </div>
  );
};

export default CoverArt;