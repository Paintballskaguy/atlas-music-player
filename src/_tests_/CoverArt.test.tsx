import { describe, it, expect, vi } from 'vitest'
import { render } from '@testing-library/react'
import CoverArt from '../components/CoverArt'
import { MusicPlayerContext } from '../contexts/MusicPlayerContext'
import React from 'react'

// Mock the useMusicPlayer hook
const mockUseMusicPlayer = vi.fn()
vi.mock('../contexts/MusicPlayerContext', () => ({
  useMusicPlayer: () => mockUseMusicPlayer(),
}))

describe('CoverArt', () => {
  const defaultMockContext = {
    currentSongDetails: null,
    lyrics: null,
    currentSong: null,
    loadLyrics: vi.fn().mockResolvedValue(undefined),
  }

  beforeEach(() => {
    mockUseMusicPlayer.mockReturnValue(defaultMockContext)
  })

  afterEach(() => {
    vi.clearAllMocks()
  })

  it('renders with default state (no cover image)', () => {
    const { container } = render(<CoverArt alt="Cover Art" />)
    
    // Should show the music note fallback
    expect(container.querySelector('.bg-custom-teal-600')).toBeInTheDocument()
    expect(container.textContent).toContain('â™ª')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders with cover image', () => {
    mockUseMusicPlayer.mockReturnValue({
      ...defaultMockContext,
      currentSongDetails: {
        id: '1',
        title: 'Test Song',
        artist: 'Test Artist',
        genre: 'Pop',
        duration: '3:00',
        cover: 'https://example.com/cover.jpg',
        song: 'https://example.com/song.mp3'
      }
    })
    
    const { container } = render(<CoverArt alt="Test Album Cover" />)
    
    // Should show the image
    const img = container.querySelector('img')
    expect(img).toHaveAttribute('src', 'https://example.com/cover.jpg')
    expect(img).toHaveAttribute('alt', 'Test Album Cover')
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders with custom alt text', () => {
    const { container } = render(<CoverArt alt="Custom Album Art" />)
    
    // The alt text should be used for the fallback (when no image)
    const img = container.querySelector('img')
    // Since there's no cover image, it will show the fallback, but the alt is still set
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders with lyrics available on hover', () => {
    mockUseMusicPlayer.mockReturnValue({
      ...defaultMockContext,
      currentSongDetails: {
        id: '1',
        title: 'Test Song',
        artist: 'Test Artist',
        genre: 'Pop',
        duration: '3:00',
        cover: 'https://example.com/cover.jpg',
        song: 'https://example.com/song.mp3'
      },
      lyrics: '[Verse 1]\nTest lyrics content',
      currentSong: { id: '1', title: 'Test Song' }
    })
    
    const { container } = render(<CoverArt alt="Test Cover" />)
    
    // Lyrics panel should be hidden initially (only shown on hover)
    const lyricsPanel = container.querySelector('.bg-black.bg-opacity-80')
    expect(lyricsPanel).not.toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('handles image error by showing fallback', () => {
    mockUseMusicPlayer.mockReturnValue({
      ...defaultMockContext,
      currentSongDetails: {
        id: '1',
        title: 'Test Song',
        artist: 'Test Artist',
        genre: 'Pop',
        duration: '3:00',
        cover: 'https://invalid-url.com/cover.jpg',
        song: 'https://example.com/song.mp3'
      }
    })
    
    const { container } = render(<CoverArt alt="Test Cover" />)
    
    // Initially tries to show image
    const img = container.querySelector('img')
    expect(img).toBeInTheDocument()
    
    // Simulate image error
    if (img) {
      img.dispatchEvent(new Event('error'))
    }
    
    // After error, should show fallback
    expect(container.firstChild).toMatchSnapshot()
  })

  it('renders with current song but no cover image', () => {
    mockUseMusicPlayer.mockReturnValue({
      ...defaultMockContext,
      currentSongDetails: null,
      currentSong: { 
        id: '1', 
        title: 'Loading Song', 
        artist: 'Loading Artist' 
      }
    })
    
    const { container } = render(<CoverArt alt="Loading Cover" />)
    
    // Should show fallback when no currentSongDetails
    expect(container.querySelector('.bg-custom-teal-600')).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })

  it('calls loadLyrics on mouse enter when current song exists', async () => {
    const mockLoadLyrics = vi.fn().mockResolvedValue(undefined)
    
    mockUseMusicPlayer.mockReturnValue({
      ...defaultMockContext,
      currentSongDetails: {
        id: '1',
        title: 'Test Song',
        artist: 'Test Artist',
        cover: 'https://example.com/cover.jpg',
        song: 'https://example.com/song.mp3'
      },
      currentSong: { id: '1', title: 'Test Song' },
      loadLyrics: mockLoadLyrics
    })
    
    const { container } = render(<CoverArt alt="Test Cover" />)
    
    // Simulate mouse enter
    const coverArtDiv = container.firstChild as HTMLElement
    coverArtDiv.dispatchEvent(new MouseEvent('mouseenter'))
    
    // Wait for async operation
    await new Promise(resolve => setTimeout(resolve, 0))
    
    expect(mockLoadLyrics).toHaveBeenCalledWith('1')
    expect(container.firstChild).toMatchSnapshot()
  })
})