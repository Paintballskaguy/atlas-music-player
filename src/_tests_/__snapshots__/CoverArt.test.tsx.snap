// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CoverArt > handles image error by showing fallback 1`] = `
<div
  class="w-full aspect-square bg-gradient-to-br from-custom-teal-100 to-custom-blue-200 dark:from-custom-teal-800 dark:to-custom-blue-900 rounded-3xl overflow-hidden flex items-center justify-center shadow-2xl border-4 border-custom-yellow-200 dark:border-custom-yellow-600 transition-all duration-300 relative cursor-pointer"
>
  <div
    class="w-full h-full flex items-center justify-center"
  >
    <img
      alt="Test Cover"
      class="w-full h-full object-cover"
      src="https://invalid-url.com/cover.jpg"
    />
  </div>
</div>
`;

exports[`CoverArt > renders with cover image 1`] = `
<div
  class="w-full aspect-square bg-gradient-to-br from-custom-teal-100 to-custom-blue-200 dark:from-custom-teal-800 dark:to-custom-blue-900 rounded-3xl overflow-hidden flex items-center justify-center shadow-2xl border-4 border-custom-yellow-200 dark:border-custom-yellow-600 transition-all duration-300 relative cursor-pointer"
>
  <div
    class="w-full h-full flex items-center justify-center"
  >
    <img
      alt="Test Album Cover"
      class="w-full h-full object-cover"
      src="https://example.com/cover.jpg"
    />
  </div>
</div>
`;

exports[`CoverArt > renders with current song but no cover image 1`] = `
<div
  class="w-full aspect-square bg-gradient-to-br from-custom-teal-100 to-custom-blue-200 dark:from-custom-teal-800 dark:to-custom-blue-900 rounded-3xl overflow-hidden flex items-center justify-center shadow-2xl border-4 border-custom-yellow-200 dark:border-custom-yellow-600 transition-all duration-300 relative cursor-pointer"
>
  <div
    class="w-full h-full flex items-center justify-center"
  >
    <div
      class="w-20 h-20 bg-custom-teal-600 dark:bg-custom-teal-300 rounded-full flex items-center justify-center"
    >
      <div
        class="text-white dark:text-custom-teal-800 text-2xl font-bold"
      >
        ♪
      </div>
    </div>
  </div>
</div>
`;

exports[`CoverArt > renders with custom alt text 1`] = `
<div
  class="w-full aspect-square bg-gradient-to-br from-custom-teal-100 to-custom-blue-200 dark:from-custom-teal-800 dark:to-custom-blue-900 rounded-3xl overflow-hidden flex items-center justify-center shadow-2xl border-4 border-custom-yellow-200 dark:border-custom-yellow-600 transition-all duration-300 relative cursor-pointer"
>
  <div
    class="w-full h-full flex items-center justify-center"
  >
    <div
      class="w-20 h-20 bg-custom-teal-600 dark:bg-custom-teal-300 rounded-full flex items-center justify-center"
    >
      <div
        class="text-white dark:text-custom-teal-800 text-2xl font-bold"
      >
        ♪
      </div>
    </div>
  </div>
</div>
`;

exports[`CoverArt > renders with default state (no cover image) 1`] = `
<div
  class="w-full aspect-square bg-gradient-to-br from-custom-teal-100 to-custom-blue-200 dark:from-custom-teal-800 dark:to-custom-blue-900 rounded-3xl overflow-hidden flex items-center justify-center shadow-2xl border-4 border-custom-yellow-200 dark:border-custom-yellow-600 transition-all duration-300 relative cursor-pointer"
>
  <div
    class="w-full h-full flex items-center justify-center"
  >
    <div
      class="w-20 h-20 bg-custom-teal-600 dark:bg-custom-teal-300 rounded-full flex items-center justify-center"
    >
      <div
        class="text-white dark:text-custom-teal-800 text-2xl font-bold"
      >
        ♪
      </div>
    </div>
  </div>
</div>
`;

exports[`CoverArt > renders with lyrics available on hover 1`] = `
<div
  class="w-full aspect-square bg-gradient-to-br from-custom-teal-100 to-custom-blue-200 dark:from-custom-teal-800 dark:to-custom-blue-900 rounded-3xl overflow-hidden flex items-center justify-center shadow-2xl border-4 border-custom-yellow-200 dark:border-custom-yellow-600 transition-all duration-300 relative cursor-pointer"
>
  <div
    class="w-full h-full flex items-center justify-center"
  >
    <img
      alt="Test Cover"
      class="w-full h-full object-cover"
      src="https://example.com/cover.jpg"
    />
  </div>
</div>
`;
